Feature/test tools: refactored 'Fitter'

jlunenburg opened the pull request:
Tried to refactor lots of stuff to improve readability and testability. Test results are the same as on master. Please pay specific attention to class, struct and variable names (in their respective scopes): is it clear what they're used for (and if no, what would be more appropriate and why)?

Before merging, it would be advisable if someone could check this in practice on the robot.

jlunenburg committed:
cleanup(fitter)Removed deprecated member variable

jlunenburg committed:
wip(fitter)Made entity shape map mutable and fitter methods const

jlunenburg committed:
refactor(test_furniture_fit)Put stuff in a class

jlunenburg committed:
refactor(fitter)Moved get2DShape to separate method

jlunenburg committed:
refactor(fitter)Moved renderWorldModel2D and computeYawRange

To separate functions

jlunenburg committed:
cleanup()Removed autosave and added gitignore

jlunenburg committed:
refactor(fitter)Moved checkExpectedBeamThroughEntity and
computeShapeCenter

jlunenburg committed:
refactor(fitter)Created transfor2DShape

jlunenburg committed:
refactor(fitter)Added computeFitterError

jlunenburg committed:
refactor(fitter)Created updateOptimum

jlunenburg committed:
refactor(fitter)Renamed some variables

jlunenburg committed:
refactor(fitter)Moved final computation to computeFittedPose

jlunenburg committed:
wip(fitter)Refactoring some more

jlunenburg committed:
cleanup(fitter)Removed some deprecated code

jlunenburg committed:
cleanup(fitter)Fixed compiler warning

jlunenburg committed:
refactor(fitter)Moved evaluating one candidate to separate function

jlunenburg committed:
refactor(fitter)Moved more functionality to the 'Candidate' class

jlunenburg committed:
refactor(fitter)Moved pre-processing to separate function

jlunenburg committed:
refactor(fitter)Implemented datastruct in EvaluateCandidate

PetervDooren commented on ed_sensor_integration/src/kinect/fitter.cpp:
```diff
@@ -47,235 +111,333 @@ geo::Transform2 XYYawToTransform2(const geo::Pose3D& pose)
+
+// ----------------------------------------------------------------------------------------------------
+
+bool Fitter::estimateEntityPoseImp(const FitterData& data, const ed::WorldModel& world, const ed::UUID& id,
+                                   const geo::Pose3D& expected_pose, geo::Pose3D& fitted_pose, double max_yaw_change) const
```
I know that adding docstrings is not the focus of this PR. But a docstring explaining this function would help a lot. There we can explain the process that is used to estimate the entitiy-pose. We can explain among other things how we determine the candidates, i.e. what we are iterating over. It can explain the definition of the yaw angle (w.r.t. what is it defined).

jlunenburg replied:
See e2ef22e71db426461225079a85e07b267bddafbc

PetervDooren commented on ed_sensor_integration/src/kinect/fitter.cpp:
```diff
@@ -47,235 +111,333 @@ geo::Transform2 XYYawToTransform2(const geo::Pose3D& pose)
     }
 
-    geo::Vec2 shape_center = 0.5 * (shape_min + shape_max);
+    double error_threshold = 1e5;
```
is there a way to relate this magic number to the one of the initial current_optimum. That would save us one magic number

jlunenburg replied:
See 8e3b620eb4aadc116630c04fb806cdf23702a7af

PetervDooren commented on ed_sensor_integration/include/ed/kinect/fitter.h:
```diff
@@ -38,43 +57,66 @@ struct FitterData
 
-    EntityRepresentation2D GetOrCreateEntity2D(const ed::EntityConstPtr& e);
+    // ToDo: make private?
+    EntityRepresentation2D GetOrCreateEntity2D(const ed::EntityConstPtr& e) const;
```
resolve your own TODO.

jlunenburg replied:
See e2ef22e71db426461225079a85e07b267bddafbc

PetervDooren commented on ed_sensor_integration/include/ed/kinect/fitter.h:
```diff
@@ -38,43 +57,66 @@ struct FitterData
+    // ToDo: make private?
     void renderEntity(const ed::EntityConstPtr& e, const geo::Pose3D& sensor_pose_xya, int identifier,
-                      std::vector<double>& model_ranges, std::vector<int>& identifiers);
+                      std::vector<double>& model_ranges, std::vector<int>& identifiers) const;
```
resolve own TODO

jlunenburg replied:
See e2ef22e71db426461225079a85e07b267bddafbc

PetervDooren commented on ed_sensor_integration/include/ed/kinect/fitter.h:
```diff
@@ -38,43 +57,66 @@ struct FitterData
 private:
 
-    // Fitting
+    bool estimateEntityPoseImp(const FitterData& data, const ed::WorldModel& world, const ed::UUID& id,
+                   const geo::Pose3D& expected_pose, geo::Pose3D& fitted_pose, double max_yaw_change) const;
```
For this one and many in this class definition:
Should this be in the header file? or can we simply put it in the code file. Especially private functions might not be needed anywhere outside this class.

jlunenburg replied:
That's true, but class functions must be declared in the class definition, hence here. If free functions would be used, you could indeed declare them in the source file. However, you wouldn't have access to the member variables in that case.

PetervDooren commented on ed_sensor_integration/src/kinect/fitter.cpp:
```diff
@@ -47,235 +111,333 @@ geo::Transform2 XYYawToTransform2(const geo::Pose3D& pose)
-        const ed::EntityConstPtr& e = *it;
-        if (e->id() == id) // Skip entity id that needs to be fitted
-            continue;
+        double ds = sensor_ranges[i];
+        double dm = test_ranges[i];
```
These represent the distance according to the sensor and model, correct? perhaps add a comment to clarify this.

jlunenburg replied:
See b6f1d89877f5c398adca0fc5e7be1c4f94a42dd5

PetervDooren commented on ed_sensor_integration/src/kinect/fitter.cpp:
```diff
@@ -47,235 +111,333 @@ geo::Transform2 XYYawToTransform2(const geo::Pose3D& pose)
 
-        if (id_str.size() >= 5 && id_str.substr(0, 5) == "amigo")
+        if (dm <= 0)
+        {
+            total_error += 0.1;
             continue;
+        }
 
-        renderEntity(e, data.sensor_pose_xya, -1, model_ranges, dummy_identifiers);
+        double diff = std::abs(ds - dm);
+        if (diff < 0.1)
+            total_error += diff;
+        else
+        {
+            if (ds > dm)
+                total_error += 1;
+            else
+                total_error += 0.1;
+        }
```
ideally we would have a lot less magic numbers in this scoring system. But I acknowledge that that is not the purpose of this PR.

jlunenburg replied:
Yes, I tried to change as little as possible content wise. Therefore, I consider this outside the scope

PetervDooren commented on ed_sensor_integration/test/test_furniture_fit.cpp:
```diff
@@ -313,58 +253,201 @@ class FurnitureFitTest : public ::testing::Test {
 
 
-TEST_F(FurnitureFitTest, testCase)
+// ToDo: make this class a member of the FurnitureFitTest and separate
+// the test functions from the test setup
+/**
+ * @brief The TestSetup class contains all constant data for the test
+ */
```
I am fine with leaving this TODO for another PR. Was this your intention?

jlunenburg replied:
Yes, this might become important once we add more tests.

PetervDooren reviewed (changes requested):
I know that it is not the focus of this PR but I think it would be a good idea to add docstrings to some ot the classes. Explain what they are and what the functions do. In case of the main function it is also a good idea to explain what the process is.
Other than that this thing is beautiful

MatthijsBurgh commented on ed_sensor_integration/src/kinect/fitter.cpp:
```diff
@@ -338,17 +312,57 @@ bool Fitter::estimateEntityPoseImp(const FitterData& data, const ed::WorldModel&
+
+    // -------------------------------------
+    // Calculate the beam which shoots through the expected position of the entity
+    geo::Vec3 expected_pos_SENSOR = data.sensor_pose_xya.inverse() * expected_pose.t;
```
Why these caps? Doesn't match any naming/case convention

jlunenburg replied:
See b6f1d89877f5c398adca0fc5e7be1c4f94a42dd5

MatthijsBurgh commented on ed_sensor_integration/include/ed/kinect/fitter.h:
```diff
@@ -83,6 +94,8 @@ class Fitter
     bool estimateEntityPoseImp(const FitterData& data, const ed::WorldModel& world, const ed::UUID& id,
                    const geo::Pose3D& expected_pose, geo::Pose3D& fitted_pose, double max_yaw_change) const;
 
+    EstimationInputData preProcessInputData(const ed::WorldModel &world, const ed::UUID &id, const geo::Pose3D &expected_pose, const FitterData &data) const; // ToDo: unique_ptr?
```
Unique_ptr to what?

jlunenburg replied:
Removed, not sure how to improve

jlunenburg committed:
cleanup(fitter)Removed some magic numbers

jlunenburg committed:
doc(Fitter)Added docstrings

jlunenburg committed:
cleanup(fitter)Minor cleanup and comments

jlunenburg commented:
I found a potential bug, need to investigate

jlunenburg committed:
fix(Fitter)Model ranges is used in fitting

This is important if the world model contains multiple entities
(including walls)

jlunenburg committed:
doc(Fitter)Added missing docstring

jlunenburg committed:
refactor(Fitter)Created 'findOptimum' function

PetervDooren reviewed (commented):
Needs to be tested on the robot. Then It is good as far as I'm concerned

jlunenburg committed:
doc(fitter)Updated some comments

PetervDooren commented:
When testing the system could not compile. It gave an error in ed_robocup, which makes use of fitter::getOrCreateEntity2D, which has been made private in this PR.

So we gotta make sure that ed_robocup is also compatible with this update

Other than that, works like a charm in the real world

MatthijsBurgh reviewed (commented):
The function @PetervDooren mentioned, should be made public again. I think that is easiest solution to keep `ed_robocup` working

PetervDooren commented:
I think we first need to wonder whether ed_robocup is abusing the function. Because this does not seem like something the fitter should be providing.

jlunenburg committed:
fix(fitter)Made GetOrCreateEntity (and fixed capital G)

The ed_robocup package makes use of this

jlunenburg commented:
> When testing the system could not compile. It gave an error in ed_robocup, which makes use of fitter::getOrCreateEntity2D, which has been made private in this PR.
> 
> So we gotta make sure that ed_robocup is also compatible with this update
> 
> Other than that, works like a charm in the real world

@PetervDooren I rolled it back and made it public again. Not sure how much work it'd be to fix it in `ed_robocup` and feel that this has taken long enough ;-)
